modules/adfinityBidAdapter.js

```diff
@@ -0,0 +1,117 @@
+
+const BIDDER_CODE = 'adfinity';
+const AD_URL = 'https://stat.adfinity.pro/?c=o&m=multi';
+const SYNC_URL = 'https://stat.adfinity.pro/?c=o&m=cookie'
+
+function isBidResponseValid(bid) {
+  if (!bid.requestId || !bid.cpm || !bid.creativeId || !bid.ttl || !bid.currency) {
+    return false;
+  }
+
+  switch (bid.mediaType) {
+    case BANNER:
+      return Boolean(bid.width && bid.height && bid.ad);
+    case VIDEO:
+      return Boolean(bid.vastUrl);
+    case NATIVE:
+      return Boolean(bid.native);
+    default:
+      return false;
+  }
+}
+
+export const spec = {
+  code: BIDDER_CODE,
+  supportedMediaTypes: [BANNER, VIDEO, NATIVE],
+  /**
+   * Determines whether or not the given bid request is valid.
+   *
+   * @param {object} bid The bid to validate.
+   * @return boolean True if this is a valid bid, and false otherwise.
+   */
+  isBidRequestValid: (bid) => {
+    return Boolean(bid.bidId && bid.params && !isNaN(bid.params.placement_id));
+  },
+
+  /**
+   * Make a server request from the list of BidRequests.
+   *
+   * @param {BidRequest[]} validBidRequests A non-empty list of valid bid requests that should be sent to the Server.
+   * @return ServerRequest Info describing the request to the server.
+   */
+  buildRequests: (validBidRequests, bidderRequest) => {
+    let winTop = window;
+    let location;
+    try {
+      location = new URL(bidderRequest.refererInfo.referer)
+      winTop = window.top;
+    } catch (e) {
+      location = winTop.location;
+      utils.logMessage(e);
+    };
+    let placements = [];
+    let request = {
+      'deviceWidth': winTop.screen.width,
+      'deviceHeight': winTop.screen.height,
+      'language': (navigator && navigator.language) ? navigator.language : '',
+      'secure': location.protocol === 'https:' ? 1 : 0,
+      'host': location.host,
+      'page': location.pathname,
+      'placements': placements
+    };
+
+    for (let i = 0; i < validBidRequests.length; i++) {
+      let bid = validBidRequests[i];
+      let traff = bid.params.traffic || BANNER
+      let placement = {
+        placementId: bid.params.placement_id,
+        bidId: bid.bidId,
+        sizes: bid.mediaTypes[traff].sizes,
+        traffic: traff
+      };
+      if (bid.schain) {
+        placement.schain = bid.schain;
+      }
+      placements.push(placement);
+    }
+    return {
+      method: 'POST',
+      url: AD_URL,
+      data: request
+    };
+  },
+
+  /**
+   * Unpack the response from the server into a list of bids.
+   *
+   * @param {*} serverResponse A successful response from the server.
+   * @return {Bid[]} An array of bids which were nested inside the server.
+   */
+  interpretResponse: (serverResponse) => {
+    let response = [];
+    try {
+      serverResponse = serverResponse.body;
+      for (let i = 0; i < serverResponse.length; i++) {
+        let resItem = serverResponse[i];
+        if (isBidResponseValid(resItem)) {
+          response.push(resItem);
+        }
```

Comment:
This current code doesn't really support native bids, at least based on the test native params that were provided.

When creating a `native` bid for Prebid, there are certain fields that need to be defined to properly identify all the normal native assets (which are used in the rendering process).  Can you update the logic here to add this support?

The following page provides some additional information if needed:
http://prebid.org/dev-docs/bidder-adaptor.html#supporting-native

Reply:
Ok, help me to understend, what prebid adapter shoukd get in response from SSP, currently, using test params we return such object
`{
                       mediaType: 'native',
                       native:{
                           clickUrl: "",
                           title: 'Test',
                           image: {
                               url: '',
                               width: 226,
                               height: 158
                           },
                           impressionTrackers: []
                       },
                       ttl: 120,
                       cpm: 0.4,
                       requestId: placement['bidId'],
                       creativeId: '2',
                       netRevenue: true,
                       currency: 'USD',
                       dealId: '1'
                   }`
This response contains minimal data to render native ad: clickUrl, impTrackers, image and title. This should be enough to test integration. To apply assets, our users stores them into ssp user UI

Reply:
So on the whole, that list you provided should be sufficient in the minimal terms.  If you're not going to add the code in the adapter to set the values for the `bidResponse`, you'll have to ensure any other potential native assets that may be used are setup properly in the SSP response that your endpoint returns.

If the values aren't formatted correctly, there's a chance that Prebid will reject that native bid as not being setup properly.  Please let me know if that makes sense.

Reply:
Building response to Prebid adapter on backend side i'm using construction to set prebid assets
like this:
`if (asset.data) {
            switch (asset.data.type) {
                case 1:
                    hbResponse.sponsoredBy = asset.data.value;
                    break;
                case 2:
                    hbResponse.body = asset.data.value;
                    break;
                case 3:
                    hbResponse.rating = asset.data.value;
                    break;
                case 4:
                    hbResponse.likes = asset.data.value;
                    break;
                case 5:
                    hbResponse.downloads = asset.data.value;
                    break;
                case 6:
                    hbResponse.price = asset.data.value;
                    break;
                case 7:
                    hbResponse.salePrice = asset.data.value;
                    break;
                case 8:
                    hbResponse.phone = asset.data.value;
                    break;
                case 9:
                    hbResponse.address = asset.data.value;
                    break;
                case 10:
                    hbResponse.body2 = asset.data.value;
                    break;
                case 11:
                    hbResponse.displayUrl = asset.data.value;
                    break;
                case 12:
                    hbResponse.cta = asset.data.value;
                    break;
                default:
                    hbResponse.body = asset.data.value;
            }
        } else if (asset.img) {
            switch (asset.img.type) {
                case 2:
                    hbResponse.icon = {
                    url: asset.img.url,
                    width: asset.img.w,
                    height: asset.img.h
                    };
                    break;
                case 3:
                    hbResponse.image = {
                    url: asset.img.url,
                    width: asset.img.w,
                    height: asset.img.h
                    };
                    break;
                default:
                    hbResponse.image = {
                        url: asset.img.url,
                        width: asset.img.w,
                        height: asset.img.h
                    };
            }
        }`